java/arcs/core/testutil/Fuzzing.kt

```diff
@@ -0,0 +1,344 @@
+  override fun nextInRange(min: Int, max: Int): Int = random.nextInt(min, max + 1)
+  override fun nextInt(): Int = random.nextInt()
+  fun printSeed() {
+    println("Seed was $seed")
+  }
+}
+
+/**
+ * A source of randomness that is seeded by the current date & time.
+ */
+class DateSeededRandom : SeededRandom(LocalDateTime.now().hashCode())
+
+/**
+ * A generator (implementation of [A<T>]) that always produces the same, specified value.
+ */
+class Value<T>(val value: T) : A<T> {
+  override operator fun invoke(): T = value
+}
+
+/**
+ * A generator (implementation of [A<T>]) that produces each value in the provided list,
+ * in order. This generator restarts from the beginning of the list when all values are
+ * exhausted.
+ */
+class SequenceOf<T>(val values: List<T>) : A<T> {
+  var idx = 0
+  override operator fun invoke(): T {
+    val result = values[idx++]
+    if (idx == values.size) {
+      idx = 0
+    }
+    return result
+  }
+}
+
+/**
+ * A generator (implementation of [A<T>]) that randomly chooses from a provided list each time
+ * it's invoked.
+ */
+class ChooseFromList<T>(val s: Seed, val values: List<T>) : A<T> {
+  override operator fun invoke(): T {
+    return values[s.nextLessThan(values.size)]
+  }
+}
+
+/**
+ * A generator (implementation of [A<List<T>>]) that randomly produces a list of the provided
+ * length, using the provided generator).
+ */
+class ListOf<T>(val generator: A<T>, val length: A<Int>) : A<List<T>> {
+  override operator fun invoke(): List<T> {
+    return (1..length()).map { generator() }
+  }
+}
+
+/**
+ * A generator (implementation of [A<Map<T, U>>]) that randomly produces a map with the provided
+ * number of entries, with keys drawn from the provided key generator and values drawn from
+ * the provided value generator.
+ */
+class MapOf<T, U>(val key: A<T>, val value: A<U>, val entries: A<Int>) : A<Map<T, U>> {
+  override operator fun invoke(): Map<T, U> {
+    val keys: MutableSet<T> = mutableSetOf<T>()
+    val map = mutableMapOf<T, U>()
+    val size = entries()
+    while (keys.size < size) {
+      keys.add(key())
+    }
+    keys.forEach {
+      map.put(it, value())
+    }
+    return map
+  }
+}
+
+/**
+ * A transformer (implementation of [T<I, O>]) that implements a specified function.
+ *
+ * Taking the following invariant:
+ * invariant_withdrawals_lessThan_initialBalance_willApply(
+ *   initalBalance: A<Int>,
+ *   withdrawal: T<Int, Withdrawal>
+ * )
+ *
+ * we can call it with an explicit withdrawal function:
+ *   invariant_withdrawals_lessThan_initialBalance_willApply(balance, Function { it/2 })
+ *
+ * or even with an explicit value:
+ *   invariant_withdrawals_lessThan_initialBalance_willApply(balance, Function { 200 })
+ */
+class Function<I, O>(val f: (i: I) -> O) : T<I, O>() {
+  override operator fun invoke(i: I): O = f(i)
+}
+
+/**
+ * Utility method to provision a fuzz test with a random seed value that's printed out
+ * if the test fails.
```

Comment:
why not always print out the seed in any case? since normally we look at the output only if the test fails anyway.

Reply:
I don't think it's good practice to print stuff into test logs when the test is passing.